pkg/services/signingkeys/signingkeystore/store.go

```diff
@@ -0,0 +1,194 @@
+		dbStore:        dbStore,
+		secretsService: secretsService,
+	}
+}
+
+func (s *Store) GetJWKS(ctx context.Context) (jose.JSONWebKeySet, error) {
+	keySet := jose.JSONWebKeySet{}
+
+	keys := []*SigningKey{}
+	err := s.dbStore.GetSqlxSession().Select(ctx, &keys, "SELECT * FROM signing_key")
+	if err != nil {
+		return keySet, err
+	}
+
+	for _, key := range keys {
+		assertedKey, err := s.decodePrivateKey(ctx, key)
+		if err != nil {
+			return keySet, err
+		}
+
+		keySet.Keys = append(keySet.Keys, jose.JSONWebKey{
+			Key:       assertedKey.Public(),
+			Algorithm: string(key.Alg),
+			KeyID:     key.KeyID,
+			Use:       "sig",
+		})
+	}
+
+	return keySet, nil
+}
+
+// AddPrivateKey adds a private key to the service.
+func (s *Store) AddPrivateKey(ctx context.Context,
+	keyID string, alg jose.SignatureAlgorithm, privateKey crypto.Signer, expiresAt *time.Time, force bool) error {
+	privateKeyPEM, err := s.encodePrivateKey(ctx, privateKey)
+	key := &SigningKey{
+		KeyID:      keyID,
+		PrivateKey: privateKeyPEM,
+		AddedAt:    time.Now(),
+		Alg:        alg,
+		ExpiresAt:  expiresAt,
+	}
+
+	dbSession := s.dbStore.GetSqlxSession()
+	err = dbSession.WithTransaction(ctx, func(tx *session.SessionTx) error {
+		existingKey := SigningKey{}
+		err := tx.Get(ctx, &existingKey, "SELECT * FROM signing_key WHERE key_id = ?", keyID)
+		if err != nil && !errors.Is(err, sql.ErrNoRows) {
+			return err
+		}
+
+		if len(existingKey.PrivateKey) == 0 {
+			_, err = tx.Exec(ctx,
+				"INSERT INTO signing_key (key_id, private_key, added_at, alg, expires_at) VALUES (?, ?, ?, ?, ?)",
+				key.KeyID, key.PrivateKey, key.AddedAt, key.Alg, key.ExpiresAt)
+			return err
+		}
+
+		if force || (existingKey.ExpiresAt != nil && existingKey.ExpiresAt.Before(time.Now())) {
+			_, err = tx.Exec(ctx,
+				"UPDATE signing_key SET private_key = ?, added_at = ?, alg = ?, expires_at = ? WHERE key_id = ?",
+				key.PrivateKey, key.AddedAt, key.Alg, key.ExpiresAt, key.KeyID)
+			return err
+		}
+
+		return nil
+	})
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// GetPrivateKey returns the private key with the specified key ID
+func (s *Store) GetPrivateKey(ctx context.Context, keyID string) (crypto.Signer, error) {
+	key := &SigningKey{}
+	err := s.dbStore.GetSqlxSession().Get(ctx, key, "SELECT * FROM signing_key WHERE key_id = ?", keyID)
+	if err != nil {
+		return nil, err
+	}
+
+	signKey, err := s.decodePrivateKey(ctx, key)
+	if err != nil {
+		return nil, err
+	}
+
+	return signKey, nil
+}
+
+func (s *Store) encodePrivateKey(ctx context.Context, privateKey crypto.Signer) ([]byte, error) {
+	// Encode private key to binary format
+	pKeyBytes, err := x509.MarshalPKCS8PrivateKey(privateKey)
+	if err != nil {
+		return nil, err
+	}
+
+	// Encode private key to PEM format
+	privateKeyPEM := pem.EncodeToMemory(&pem.Block{
+		Type:  "PRIVATE KEY",
+		Bytes: pKeyBytes,
+	})
+
+	encrypted, err := s.secretsService.Encrypt(ctx, privateKeyPEM, secrets.WithoutScope())
```

Comment:
which encryption key is behind `secrets.WithoutScope()` ?

Reply:
the `root` key, same one used by oauth token encryption and secret kvstore. 

It's either generated on the fly or uses a different secrets backend from my fast research

https://github.com/grafana/grafana/blob/main/pkg/services/secrets/manager/manager.go#L187